kevm-pyk/src/kevm_pyk/kevm.py

```diff
@@ -267,6 +267,17 @@ def is_terminal(cterm: CTerm) -> bool:
                     return True
         return False
 
+    @staticmethod
+    def same_loop(cterm1: CTerm, cterm2: CTerm) -> bool:
+        branch1 = list(KEVM.extract_branches(cterm1))
+        branch2 = list(KEVM.extract_branches(cterm2))
+        same_branch = len(branch1) > 0 and len(branch2) > 0 and branch1[0] == branch2[0]
+        same_cell_structure = all(
+            cterm1.cell(cn) == cterm2.cell(cn) for cn in ['PC_CELL', 'CALLDATA_CELL', 'PROGRAM_CELL', 'JUMPDESTS_CELL']
+        )
+        same_wordstack_structure = KEVM.wordstack_len(cterm1) == KEVM.wordstack_len(cterm2)
+        return same_branch and same_cell_structure and same_wordstack_structure
+
```

Comment:
This probably isn't doing what you expect, and will fail for any loops that are not very simple.

Here: https://github.com/runtimeverification/ksummarize/blob/master/ksummarize/src/ksummarize/evm_strategy.py#L156, we are using method `_extract_branch`, which returns a tuple of `pcount, cond` (program counter and condition) for the _pre-branch_ node, but you are using `KEVM.extract_branches` (which is implemented in the above link as `extract_branch_condition`, which returns the actual branch conditions.

So in the original, we are comparing that the program counter being jumped to (the first argument of `JUMPI`) is the same in both, but here you are comparing that the branching condition itself is exactly the same (which is likely not to be the case for anything but trivial loops).

Likely, what you need to do, is something more like:

```suggestion
    @staticmethod
    def same_loop(cterm1: CTerm, cterm2: CTerm) -> bool:
        # In the same program, at the same calldepth, at the same program counter
        for cell in ['PC_CELL', 'CALLDEPTH_CELL', 'PROGRAM_CELL']:
            if cterm1.cell(cell) != cterm2.cell(cell):
                return False
        # duplicate from KEVM.extract_branches
        jumpi_pattern = KEVM.jumpi_applied(KVariable('###PCOUNT'), KVariable('###COND'))
        pc_next_pattern = KApply('#pc[_]_EVM_InternalOp_OpCode', [KEVM.jumpi()])
        branch_pattern = KSequence([jumpi_pattern, pc_next_pattern, KEVM.sharp_execute(), KVariable('###CONTINUATION')])
        subst1 = branch_pattern.match(cterm1.cell('K_CELL'))
        subst2 = branch_pattern.match(cterm2.cell('K_CELL'))
        # Jumping to the same program counter
        if subst1 is not None and subst2 is not None and subst1['###PCOUNT'] == subst2['###PCOUNT']:
            if KEVM.wordstack_len(cterm1.cell('WORDSTACK_CELL')) == KEVM.wordstack_len(cterm2.cell('WORDSTACK_CELL')):
                return True
        return False
```

Note that above I've assumed that we pass a `KInner` into `KEVM.wordstack_len`.